RecoMET/METPUSubtraction/plugins/DeepMETPVRobustProducer.cc

```diff
@@ -0,0 +1,169 @@
+
+  // static methods for handling the global cache
+  static std::unique_ptr<DeepMETPVRobustCache> initializeGlobalCache(const edm::ParameterSet&);
+  static void globalEndJob(DeepMETPVRobustCache*);
+
+private:
+  const edm::EDGetTokenT<std::vector<pat::PackedCandidate> > pf_token_;
+  const bool usePUPPI_;
+  const float norm_;
+  const bool ignore_leptons_;
+  const unsigned int max_n_pf_;
+  const bool do_print_;
+
+  tensorflow::Session* session_;
+
+  tensorflow::Tensor input_;
+  tensorflow::Tensor input_cat0_;
+  tensorflow::Tensor input_cat1_;
+
+  inline static const std::unordered_map<int, int32_t> charge_embedding_{{-1, 0}, {0, 1}, {1, 2}};
+  inline static const std::unordered_map<int, int32_t> pdg_id_embedding_{
+      {-211, 0}, {-13, 1}, {-11, 2}, {0, 3}, {1, 4}, {2, 5}, {11, 6}, {13, 7}, {22, 8}, {130, 9}, {211, 10}};
+};
+
+namespace {
+  float scale_and_rm_outlier(float val, float scale) {
+    float ret_val = val * scale;
+    if (ret_val > 1e6 || ret_val < -1e6)
+      return 0.;
+    return ret_val;
+  }
+}  // namespace
+
+DeepMETPVRobustProducer::DeepMETPVRobustProducer(const edm::ParameterSet& cfg, const DeepMETPVRobustCache* cache)
+    : pf_token_(consumes<std::vector<pat::PackedCandidate> >(cfg.getParameter<edm::InputTag>("pf_src"))),
+      usePUPPI_(cfg.getParameter<bool>("usePUPPI")),
+      norm_(cfg.getParameter<double>("norm_factor")),
+      ignore_leptons_(cfg.getParameter<bool>("ignore_leptons")),
+      max_n_pf_(cfg.getParameter<unsigned int>("max_n_pf")),
+      do_print_(cfg.getParameter<bool>("do_print")),
+      session_(tensorflow::createSession(cache->graph_def)) {
+  produces<pat::METCollection>();
+
+  const int n_input = usePUPPI_ ? 7: 6;
+  const tensorflow::TensorShape shape({1, max_n_pf_, n_input});
+  const tensorflow::TensorShape cat_shape({1, max_n_pf_, 1});
+
+  input_ = tensorflow::Tensor(tensorflow::DT_FLOAT, shape);
+  input_cat0_ = tensorflow::Tensor(tensorflow::DT_FLOAT, cat_shape);
+  input_cat1_ = tensorflow::Tensor(tensorflow::DT_FLOAT, cat_shape);
+}
+
+void DeepMETPVRobustProducer::produce(edm::Event& event, const edm::EventSetup& setup) {
+  auto const& pfs = event.get(pf_token_);
+
+  const tensorflow::NamedTensorList input_list = {
+      {"input", input_}, {"input_cat0", input_cat0_}, {"input_cat1", input_cat1_}};
+
+  // Set all inputs to zero
+  input_.flat<float>().setZero();
+  input_cat0_.flat<float>().setZero();
+  input_cat1_.flat<float>().setZero();
+
+  size_t i_pf = 0;
+  float px_leptons = 0.;
+  float py_leptons = 0.;
+  const float scale = 1. / norm_;
+  for (const auto& pf : pfs) {
+    if (ignore_leptons_) {
+      int pdg_id = std::abs(pf.pdgId());
+      if (pdg_id == 11 || pdg_id == 13) {
+        px_leptons += pf.px();
+        py_leptons += pf.py();
+        continue;
+      }
+    }
+
+    // fill the tensor
+    // dz, eta, mass, pt, puppi, px, py
+    float* ptr = &input_.tensor<float, 3>()(0, i_pf, 0);
+    *ptr = pf.dz();
```

Comment:
Doesn't this just compute delta z with respect to the first primary vertex in the collection? (or is that already set to the one closest to the muon at this stage?)

Reply:
working on propogating the updated information to the DeepMET producer currently (updates soon).

maybe take a look at https://github.com/WMass/cmssw/pull/33/files#diff-9eb7cb44e20caecc7eb7b328c9c2ee434995e8991a6f07d5eb231417b9ad50c5R97-R127 this part first and see if the PV determinition looks fine?